コメントを付けられるようにするの課題

ham-cap opened the pull request:
「コメントを付けられるようにする」の課題を提出します。
よろしくお願いいたします。

ham-cap committed:
ERB Lintを導入・実行し、offenceをオートコレクトで修正

・RailsのプラクティスでERB Lintの実行が必須になったため導入し、動作確認のため実行したところoffenceが複数でたためオートコレクトで修正した。

ham-cap committed:
日報のCRUD機能を実装

ham-cap committed:
reportとbookとcommentをポリモーフィック関連付けした

ham-cap committed:
コメントの新規投稿機能を実装

ham-cap committed:
本に対するコメント投稿機能の処理を一部修正・日報に対するコメント投稿機能を実装・本と日報の詳細画面にコメント一覧を表示・日報とコメントに関わる部分の日本語化

・controllerで空のコメントインスタンスをbuildして本の詳細画面のviewに渡していたが、viewの中でコメント一覧を取得するとその空のインスタンスが混じってしまい、プロパティを呼び出す際にエラーが出るためbuildするタイミングを変更した

ham-cap committed:
コメント本文を入力必須とした

ham-cap committed:
コメントの編集機能は未実装のためEditのリンクを削除・commentable_typeによってBackのリンク先を分岐させた

ham-cap committed:
不要なファイルを削除・rubocopのoffenceを修正

yuuu commented on app/controllers/books_controller.rb:
```diff
@@ -11,7 +11,10 @@ def index
-  def show; end
+  def show
+    @book = Book.find(params[:id])
+    @comments = @book.comments.includes(:user)
```
N+1問題の対策ができていて良いですね 👍

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,68 @@
+
+  # GET /comments or /comments.json
+  def index
+    @comments = Comment.all
```
このアクションはおそらく未使用のはずなので、削除しておきましょう 🙏 
show, create以外は全て未使用に見えます 👀

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,75 @@
+  def create
+    @report = Report.new(report_params)
+
+    respond_to do |format|
```
今回の用途だとJSONをレスポンスすることはないはずなので、 `respond_to` は削除しましょう。併せてコメントの `〜.json` についても削除をお願いいたします

yuuu commented on app/views/reports/_form.html.erb:
```diff
@@ -0,0 +1,31 @@
+  </div>
+
+  <div class="field">
+    <%= form.hidden_field :user_id, :value => @user_id %>
```
これはブラウザの開発者ツールなどを使用して、ユーザー自身で書き換えができてしまうので、他人になりすまして投稿ができてしまいますね。`ReportsController#create` にて `current_user` と紐付けるようにしましょう。

ham-cap replied:
この点、なりすましの具体的な方法としては、開発者ツール上で`@user_id`の部分を任意のidに書き換えて、その状態で`登録する`のボタンを押して送信すれば投稿ができてしまうという理解で合っていますか？👀

yuuu replied:
はい。そのとおりです。

yuuu commented on app/views/reports/edit.html.erb:
```diff
@@ -0,0 +1,6 @@
+
+<%= render 'form', report: @report %>
+
+<%= link_to 'Show', @report %> |
```
ReportsのViewについて全体的にi18nの対応が漏れているようなので修正をお願いいたします 🙏

yuuu commented on app/views/reports/show.html.erb:
```diff
@@ -0,0 +1,44 @@
+          <td><%= comment.user.email %></td>
+        <% end %>
+        <td><%= comment.content %></td>
+        <td><%= comment.created_at.strftime("%Y年%-m月%-d日%H時%M分") %></td>
```
`strftime` を使うと日本における時刻表記に固定化されてしまうので、`I18n.l` を使った方が良いですね。
https://railsguides.jp/i18n.html#%E6%97%A5%E4%BB%98%E3%83%BB%E6%99%82%E5%88%BB%E3%83%95%E3%82%A9%E3%83%BC%E3%83%9E%E3%83%83%E3%83%88%E3%82%92%E8%BF%BD%E5%8A%A0%E3%81%99%E3%82%8B

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,68 @@
+    @comment = @commentable.comments.new(user_id: current_user.id, content: comment_params[:content])
+    respond_to do |format|
+      if @comment.save
+        format.html { redirect_to [@commentable, @comment], notice: 'Comment was successfully created.' }
```
`notice` のi18n対応が漏れているようなので、修正をお願いいたします

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,68 @@
+      if @comment.save
+        format.html { redirect_to [@commentable, @comment], notice: 'Comment was successfully created.' }
+      else
+        format.html { render :new, status: :unprocessable_entity }
```
`comments/new` というViewファイルはなさそうなので、実際にバリデーションエラーが起こると、ここでエラーが起こる気がします。

ham-cap committed:
不要なアクションとjsonレスポンス部分を削除・noticeをi18n対応

・comments_controllerではcreateとshow以外のアクションを使用しないため削除した
・今回はjsonのレスポンスはしないため該当部分は削除した

ham-cap committed:
commentとreportのviewでi18n対応が抜けていた部分を修正

ham-cap committed:
reportの投稿者をviewではなくcontroller側で紐付けるように変更

・reportの投稿者のidはhidden_fieldを用いてviewからcontrollerへ送っていたが、なりすまし等セキュリティ上の懸念があるためcontroller側でcurrent_userと紐付けるように変更した

yuuu commented on app/views/comments/show.html.erb:
```diff
@@ -1,11 +1,11 @@
 </p>
 <% if @comment.commentable_type == "Book" %>
-  <%= link_to 'Back', book_path(@comment.commentable_id) %>
+  <%= link_to t('views.common.back'), book_path(@comment.commentable_id) %>
```
前回レビュー時は気づかなかったのですが、if-elseでコードが重複しているので、第2引数のみを切り替えるようにしてはどうでしょうか？

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,67 @@
+  # GET /reports/new
+  def new
+    @report = Report.new
+    @user_id = current_user.id
```
`@user_id` は不要となったはずなので削除をお願いいたします。 `edit` も同様です。

yuuu commented on app/views/books/show.html.erb:
```diff
@@ -20,5 +20,31 @@
+  <tbody>
+    <% @comments.each do |comment| %>
+      <tr>
+        <% if comment.user.name %>
```
ここは他のViewでも使用しうるロジックなので、 `user.rb` に記載すると良いですね。

yuuu commented on app/views/books/show.html.erb:
```diff
@@ -20,5 +20,31 @@
+          <td><%= comment.user.email %></td>
+        <% end %>
+        <td><%= comment.content %></td>
+        <td><%= comment.created_at.strftime("%Y年%-m月%-d日%H時%M分") %></td>
```
ここもI18n.lを使いましょう。

ham-cap committed:
reports_controllerのnewとeditアクションから@user_idをセットする部分を削除

ham-cap committed:
comments/show.html.erbでコードの重複を解消

・if文の中で大部分のコードが重複していたため、重複していない部分のみを切り替えることで重複をなくした

ham-cap committed:
userの名前かメールアドレスを表示する部分をメソッドとしてuserモデル内に切り出した・コメントの投稿日時をi18n対応

・commentに紐付いたuserの名前かメールアドレスを表示する処理をif文でviewに書いていたが、別のviewでも使えそうなものだったためuserのインスタンスメソッド化した

ham-cap commented:
ご指摘いただいた部分を修正してみましたのでご確認をお願いいたします🙏

yuuu commented on app/models/user.rb:
```diff
@@ -32,4 +32,12 @@ def unfollow(user)
   end
+
+  def display_name_or_email(comment)
+    if comment.user.name
```
ここは `comment.user.name || comment.user.email` と書けますね。

yuuu commented on app/views/comments/show.html.erb:
```diff
@@ -4,8 +4,7 @@
-<% end %>
+
+<p>
+<%= link_to t('views.common.back'), @commentable_path %>
```
ここは自分のコメントも分かりづらかったかもしれません 🙏 コメントに応じてリンクを切り替えるという、見た目に関する処理がコントローラに移動してしまったので、ここはViewに寄せた方がいいと思いました。


第2引数のみを三項演算子で切り替えるとスッキリすると思うのですがどうでしょうか？

```erb
<%= link_to t('views.common.back'), @comment.commentable.is_a?(Book) ? book_path(@comment.commentable_id) : report_path(@comment.commentable_id) %>
```

今回のケースだと、以下のようにしても機能するかもしれません。

```erb
<%= link_to t('views.common.back'), @comment.commentable %>
```

ham-cap committed:
display_name_or_email内の処理を変更

ham-cap committed:
戻るボタンのリンク先の設定をcontrollerではなくviewで行うようにした
